## KoBERT 불러오기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pip install mxnet
!pip install gluonnlp pandas tqdm
!pip install sentencepiece
!pip install transformers==3.0.2
!pip install torch

  Using cached transformers-3.0.2-py3-none-any.whl (769 kB)
  Using cached tokenizers-0.8.1rc1-cp37-cp37m-manylinux1_x86_64.whl (3.0 MB)
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.11.6
    Uninstalling tokenizers-0.11.6:
      Successfully uninstalled tokenizers-0.11.6
  Attempting uninstall: transformers
    Found existing installation: transformers 4.17.0
    Uninstalling transformers-4.17.0:
      Successfully uninstalled transformers-4.17.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
kobert 0.2.3 requires transformers>=4.8.1, but you have transformers 3.0.2 which is incompatible.


In [ ]:
!pip install git+https://git@github.com/SKTBrain/KoBERT.git@master

  Cloning https://****@github.com/SKTBrain/KoBERT.git (to revision master) to /tmp/pip-req-build-bn8q74md
  Running command git clone -q 'https://****@github.com/SKTBrain/KoBERT.git' /tmp/pip-req-build-bn8q74md
  Using cached transformers-4.17.0-py3-none-any.whl (3.8 MB)
  Using cached tokenizers-0.11.6-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (6.5 MB)
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.8.1rc1
    Uninstalling tokenizers-0.8.1rc1:
      Successfully uninstalled tokenizers-0.8.1rc1
  Attempting uninstall: transformers
    Found existing installation: transformers 3.0.2
    Uninstalling transformers-3.0.2:
      Successfully uninstalled transformers-3.0.2


In [ ]:
import torch
from torch import nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import gluonnlp as nlp
import numpy as np
from tqdm import tqdm, tqdm_notebook

In [ ]:
#kobert
from kobert.utils import get_tokenizer
from kobert.pytorch_kobert import get_pytorch_kobert_model

#transformers
from transformers import AdamW
from transformers.optimization import get_cosine_schedule_with_warmup

In [ ]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)

cuda:0


In [ ]:
bertmodel, vocab = get_pytorch_kobert_model()

using cached model. /content/.cache/kobert_v1.zip
using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


## 데이터 가져오기

In [ ]:
import numpy as np
import pandas as pd
import os

In [ ]:
review = pd.read_csv("/content/drive/Shareddrives/NLP모델링/reviews_label.csv")

In [ ]:
review.head()

Unnamed: 0           학정번호      강의명  교수명        유의사항   별점     수강시기  \
0          15  YCC1002-01-00  대학기본영어Ⅱ  오진숙  대면강의, 절대평가  5.0  17년 2학기   
1          16  YCC1002-01-00  대학기본영어Ⅱ  오진숙  대면강의, 절대평가  5.0  19년 2학기   
2          17  YCC1002-01-00  대학기본영어Ⅱ  오진숙  대면강의, 절대평가  5.0  19년 2학기   
3          18  YCC1002-01-00  대학기본영어Ⅱ  오진숙  대면강의, 절대평가  4.0  19년 2학기   
4          19  YCC1002-01-00  대학기본영어Ⅱ  오진숙  대면강의, 절대평가  5.0  16년 2학기   

                                                 강의평      강의명/교수명   성적   로드  \
0                   어렵지 않은 내용으로 탄탄한 기본기를 다져주는 유익한 수업  대학기본영어Ⅱ/오진숙 -1.0 -1.0   
1  교수님 너무 귀여우심... 사랑해요\n자잘한 과제들 귀찮긴 했지만 시간투자 그렇게 ...  대학기본영어Ⅱ/오진숙 -1.0  0.5   
2  완전 좋음 2학기 때 내가 살아있던 이유.. 기말 때 갑자기 시험을 세번 본 거는 ...  대학기본영어Ⅱ/오진숙  1.0  0.5   
3  영어 말하는 데에 자신감을 얻고 싶으면 추천 다만 과제는 자잘한게 많고 필자는 1학...  대학기본영어Ⅱ/오진숙  1.0  0.5   
4  수강당시에는 조모임 비율도 높았고 발음 중시하시는건 왜 그러시는지 잘 이해가지 않지...  대학기본영어Ⅱ/오진숙  1.0  0.5   

   강의력  교수님성향  강의자료질  족보여부  
0  0.5   -1.0   -1.0  -1.0  
1  0.5    1.0   -1.0  -1.0  
2  0.5    1.0   -1.0  -1.0  
3  0.5    1.0   -1.0  -1.0  
4  0.5    1.0   -1.0  -1.0

In [ ]:
col_name = ['score_review','load_review','teaching_review','teacher_review','quality_review','jokbo' ]

for i, col in enumerate(['성적','로드','강의력','교수님성향','강의자료질','족보여부']):
  temp = review[['학정번호','강의명','강의평',col]]
  temp = temp[temp[col] != -1]
  temp['성적'] = temp[col].apply(lambda x : int(float(x)*2))
  col_name[i] = temp


In [ ]:
for col in col_name:
  print(col.iloc[:,-1].unique())

[2 1 0]
[1 0 2]
[1 0 2]
[2 1 0]
[1 0 2]
[0 1 2]


In [ ]:
dlist = ['score_list','load_list','teaching_list','teacher_list','quality_list','jokbo_list']
for i, col in enumerate(col_name):
  data_list = []
  
  for q, label in zip(col['강의평'], col.iloc[:,-1])  :
      data = []
      data.append(q)
      data.append(str(label))

      data_list.append(data)
   
  dlist[i] = data_list   

In [ ]:
for lst in dlist:
  print(len(lst))
print('\n')
for lst in dlist:
  print(lst[100])

18672
17935
20606
14234
4877
1522


['귀찮은게 많은데 다 하면 학점 잘주세요 학기중엔 싫다가 학점받고 좋아지는 교수님', '1']
['그냥 과제 앵간히 자잘히 많고 시험 앵간히 어려운 수업. 꿀강은 절대 아니다. 굳이 이 교수님 꺼 들으라고 추천은 하고 싶지 않다.\n시험은 완전 고등학교 내신 식으로 단어 빈칸 채우기 같은 사소한 느낌의 문제가 나오는데 생각보다 까다롭고, 과제의 양 자체는 막 많지는 않지만 교수님이 원하시는 규격이 있어서 그걸 무조건 맞춰야 통과된다. 아니면 무서운 인상을 쓰시며 빠꾸놓으신다.. 교수님이 좀 무서우셔서 쿠크다스 심장을 가진 사람들은 종종 심장이 바사삭될지도 ㅎㅎㅎ 아 그리고 지각이랑 결석 엄청 싫어하신다.\n결론은 고등학교 때처럼 수업에 집중해서 발표 좀 하고 수업 다 출석하고 과제 늦지 않게 내고 내신공부하듯 공부하면 좋은 성적 받을 수 있다! 는 것인데 필자는 그렇지 못해서 b권대를 받았다.. ㅠ', '1']
['깐깐한 교수님인건 맞지만 보면 볼수록 학생들의 편의를 봐주시려고 노력하는 편입니다.\n과제도 그렇게 많지 않고 학점도 잘주셔서(아마 쁠몰인듯) 무난한 강의였던 것 같습니다.', '0']
['발음좋으면 교수님이 좋아해줘요....ㅎ\n자신이 영어를 잘한다 생각해도, 항상 배울점은 많고 자신이 많이 부족하다는 점을 느끼게 해주는 강의라고 생각합니다', '2']
['일단 교수님이 매우 스윗하십니다,,,,ㅎㅎ\n정말 천사세요,,,,,\n\n중간, 기말 모두 보고서 대체인데 배우는 내용도 많이 어렵지 않아서\n수업 교재랑 교수님 피피티만 잘 참고하면 쉽게 쓸 수 있습니다.\n성적도 잘 주십니다. 수신 때 증원 많이 해주시니깐 그냥 무조건 담아두세요!', '1']
['강의가 광장히 깔끔하고 강의력이 좋으셨어요! 사실 4주차부터는 출석채우려고 틀어놓기만 하고 안들었어요 ㅋㅋㅋㅋ 그래도 과제 기간안에 분량맞춰 내고 기말 키워드 시험 3~4문제 빼고 맞추니까 a0나왔네요.\n교재랑 족보는 필수니까 꼭 

In [ ]:
#train & test 데이터로 나누기
from sklearn.model_selection import train_test_split

tr_list = ['score_train','load_train','teaching_train','teacher_train','quality_train','jokbo_train']
te_list = ['score_test','load_test','teaching_test','teacher_test','quality_test','jokbo_test']

for i, lst in enumerate(dlist):
  dataset_train, dataset_test = train_test_split(lst, test_size=0.20, random_state=0)
  tr_list[i] = dataset_train
  te_list[i] = dataset_test


In [ ]:
for lst in tr_list:
  print(len(lst))

print('\n')

for lst in te_list:
  print(len(lst))

14937
14348
16484
11387
3901
1217


3735
3587
4122
2847
976
305


# 나눠서 학습시켜보기

### 성적

In [ ]:
dataset_train = tr_list[0]
dataset_test = te_list[0]

len(dataset_train), len(dataset_test)

(14937, 3735)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 64
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2, 4368,  517, 5630, 5488, 6853, 7139, 4948, 7220, 3461, 3942,
        4213, 6705, 5495, 3005, 3432, 2462, 7101, 3942, 4444, 6150, 4939,
         517, 5906,    3,    1,    1,    1,    1,    1,    1,    1,    1,
           1,    1,    1,    1,    1,    1,    1,    1,    1,    1,    1,
           1,    1,    1,    1,    1,    1,    1,    1,    1,    1,    1,
           1,    1,    1,    1,    1,    1,    1,    1,    1], dtype=int32),
 array(25, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       dtype=int32),
 2)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/score(5)_0402.h5').to(device)

In [ ]:
#BERT 모델 불러오기
#model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/234 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 0.4330192506313324 train acc 0.84375


KeyboardInterrupt: ignored

In [ ]:
test_acc = 0.0

model.eval()
for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

In [ ]:
from sklearn.metrics import classification_report

target_names = ['깐깐함', '학점느님']

print(classification_report(label.cpu().numpy(), max_indices.cpu().numpy(), target_names=target_names))

In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/score(10)_0404.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/score(10)_0404.h5')

### 로드

In [ ]:
dataset_train = tr_list[1]
dataset_test = te_list[1]

len(dataset_train), len(dataset_test)

(14348, 3587)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 64
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2,  520, 1321, 6629,  520,  517,   46, 2872, 5350, 6736, 1259,
        6896, 1231, 6779, 5561, 1107, 6204, 6903, 3990, 2801, 5547, 7158,
        2355, 6828, 6198, 4244, 5760, 4955, 7096, 3854, 5840, 3030, 7850,
        5760, 3135, 5900,  517,   54,  517, 7930, 6705,  517, 6128, 6896,
        3443, 5740, 5784, 1628, 1444, 7147,  517,   54, 1995, 2485, 6255,
        6667, 3803, 6079, 1185, 2890, 7088, 1799, 5400,    3], dtype=int32),
 array(64, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       dtype=int32),
 0)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
#model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)
model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/load(3)_0402.h5').to(device)
#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/225 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 0.7308348417282104 train acc 0.71875
epoch 1 batch id 201 loss 0.8206743001937866 train acc 0.6958955223880597
epoch 1 train acc 0.6935416666666666


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/57 [00:00<?, ?it/s]

epoch 1 test acc 0.6386147660818713


  0%|          | 0/225 [00:00<?, ?it/s]

epoch 2 batch id 1 loss 0.7400019764900208 train acc 0.71875
epoch 2 batch id 201 loss 0.6998628377914429 train acc 0.7144745024875622
epoch 2 train acc 0.7162731481481482


  0%|          | 0/57 [00:00<?, ?it/s]

epoch 2 test acc 0.6573464912280702


  0%|          | 0/225 [00:00<?, ?it/s]

epoch 3 batch id 1 loss 0.5982980132102966 train acc 0.75
epoch 3 batch id 201 loss 0.5781843662261963 train acc 0.7755752487562189
epoch 3 train acc 0.7745833333333333


  0%|          | 0/57 [00:00<?, ?it/s]

epoch 3 test acc 0.6329495614035088


  0%|          | 0/225 [00:00<?, ?it/s]

epoch 4 batch id 1 loss 0.4551992416381836 train acc 0.796875
epoch 4 batch id 201 loss 0.4871004521846771 train acc 0.832089552238806
epoch 4 train acc 0.8311111111111111


  0%|          | 0/57 [00:00<?, ?it/s]

epoch 4 test acc 0.6699561403508771


  0%|          | 0/225 [00:00<?, ?it/s]

epoch 5 batch id 1 loss 0.2889763414859772 train acc 0.875
epoch 5 batch id 201 loss 0.44793128967285156 train acc 0.8760105721393034
epoch 5 train acc 0.8739583333333333


  0%|          | 0/57 [00:00<?, ?it/s]

epoch 5 test acc 0.6699561403508771


In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/load(8)_0404.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/load(8)_0404.h5')

### 강의력

In [ ]:
dataset_train = tr_list[2]
dataset_test = te_list[2]

len(dataset_train), len(dataset_test)

(16484, 4122)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 64
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2,  529,  517,   54,  807, 7095, 1449,  525,  525,  517, 5330,
        6359, 5439, 3969, 6269, 7089,  517,   54, 1529, 1799, 5760, 1661,
        7090, 7088, 4948, 6234, 7203, 2193, 6369, 2961, 7088, 4130, 7848,
        7288, 5782, 5760, 4078, 4193, 3925, 6896,  517, 7095, 7095, 5330,
        3860,  807, 7095, 5330, 3093, 5777, 5591, 4982,  517,   54, 4197,
        7268,  517, 5441, 2633, 6238,  517,   46,  517,    3], dtype=int32),
 array(64, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       dtype=int32),
 2)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/258 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 0.6072567701339722 train acc 0.796875
epoch 1 batch id 201 loss 0.7071292996406555 train acc 0.7014925373134329
epoch 1 train acc 0.7031519164513351


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/65 [00:00<?, ?it/s]

epoch 1 test acc 0.6245007396449704


  0%|          | 0/258 [00:00<?, ?it/s]

epoch 2 batch id 1 loss 0.722794234752655 train acc 0.6875
epoch 2 batch id 201 loss 0.45439040660858154 train acc 0.742148631840796
epoch 2 train acc 0.7537615740740741


  0%|          | 0/65 [00:00<?, ?it/s]

epoch 2 test acc 0.6422522189349112


  0%|          | 0/258 [00:00<?, ?it/s]

epoch 3 batch id 1 loss 0.4698921740055084 train acc 0.859375
epoch 3 batch id 201 loss 0.3226555585861206 train acc 0.8180192786069652
epoch 3 train acc 0.8260928079242034


  0%|          | 0/65 [00:00<?, ?it/s]

epoch 3 test acc 0.6473187869822485


  0%|          | 0/258 [00:00<?, ?it/s]

epoch 4 batch id 1 loss 0.3370414674282074 train acc 0.875
epoch 4 batch id 201 loss 0.27240267395973206 train acc 0.8790422885572139
epoch 4 train acc 0.8826240848406546


  0%|          | 0/65 [00:00<?, ?it/s]

epoch 4 test acc 0.6470784023668639


  0%|          | 0/258 [00:00<?, ?it/s]

epoch 5 batch id 1 loss 0.3001718521118164 train acc 0.921875
epoch 5 batch id 201 loss 0.18050171434879303 train acc 0.9079601990049752
epoch 5 train acc 0.9121379737295435


  0%|          | 0/65 [00:00<?, ?it/s]

epoch 5 test acc 0.6508136094674556


In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/teaching(8)_0404.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/teaching(8)_0404.h5')

### 교수님 성격

In [ ]:
dataset_train = tr_list[3]
dataset_test = te_list[3]

len(dataset_train), len(dataset_test)

(11387, 2847)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 64
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2, 3005, 7086, 3271, 4258,  525, 1258, 6160, 2359, 6079, 1665,
        5903,  517,   54, 2095, 5663, 7828, 1103, 6853,  517,   54, 4164,
        6150, 3367, 4939,  635,  517,   45, 2233, 2872, 3883,  517,   54,
        1107, 5778, 1458, 4420, 7782, 6705, 5439, 4162, 6213, 7136, 1278,
        2890, 7096, 5561, 1849,  517, 7258, 5932, 4626, 7848, 7350, 2872,
        3883,  517,   54, 2923, 6872, 6653, 6872, 1799,    3], dtype=int32),
 array(64, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       dtype=int32),
 2)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/178 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 0.42209357023239136 train acc 0.84375
epoch 1 train acc 0.8375755808417444


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/45 [00:00<?, ?it/s]

epoch 1 test acc 0.7650985663082437


  0%|          | 0/178 [00:00<?, ?it/s]

epoch 2 batch id 1 loss 0.5479536652565002 train acc 0.8125
epoch 2 train acc 0.8685622381451152


  0%|          | 0/45 [00:00<?, ?it/s]

epoch 2 test acc 0.7703068996415771


  0%|          | 0/178 [00:00<?, ?it/s]

epoch 3 batch id 1 loss 0.4198031425476074 train acc 0.8125
epoch 3 train acc 0.9107045086650162


  0%|          | 0/45 [00:00<?, ?it/s]

epoch 3 test acc 0.7758624551971326


  0%|          | 0/178 [00:00<?, ?it/s]

epoch 4 batch id 1 loss 0.34798485040664673 train acc 0.875
epoch 4 train acc 0.9420497286231195


  0%|          | 0/45 [00:00<?, ?it/s]

epoch 4 test acc 0.7786402329749104


  0%|          | 0/178 [00:00<?, ?it/s]

epoch 5 batch id 1 loss 0.29369574785232544 train acc 0.90625
epoch 5 train acc 0.9589185393258427


  0%|          | 0/45 [00:00<?, ?it/s]

epoch 5 test acc 0.7723902329749104


In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/teacher(8)_0404.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/teacher(8)_0404.h5')

### 강의질

In [ ]:
import gc
gc.collect()
torch.cuda.empty_cache()

In [ ]:
dataset_train = tr_list[4]
dataset_test = te_list[4]

len(dataset_train), len(dataset_test)

(3901, 976)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 32
batch_size = 32
warmup_ratio = 0.1
num_epochs = 10
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2, 1107, 5778, 1174, 6916, 7005, 6745,  517,   54, 3945, 1342,
        6004,  517, 5330, 5646,  835, 5920, 6137, 7088, 4617, 6705, 5761,
        4468, 6909, 4396, 7316, 5655, 7330, 3757, 6642, 7848,    3],
       dtype=int32),
 array(32, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32),
 1)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/122 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 1.0892906188964844 train acc 0.34375
epoch 1 train acc 0.7486397682306388


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 1 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 2 batch id 1 loss 0.6526179909706116 train acc 0.8125
epoch 2 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 2 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 3 batch id 1 loss 0.6507765054702759 train acc 0.8125
epoch 3 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 3 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 4 batch id 1 loss 0.6752448678016663 train acc 0.8125
epoch 4 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 4 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 5 batch id 1 loss 0.6468477845191956 train acc 0.8125
epoch 5 train acc 0.8003815715093273


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 5 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 6 batch id 1 loss 0.6568831205368042 train acc 0.8125
epoch 6 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 6 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 7 batch id 1 loss 0.6420496702194214 train acc 0.8125
epoch 7 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 7 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 8 batch id 1 loss 0.6334251761436462 train acc 0.8125
epoch 8 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 8 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 9 batch id 1 loss 0.619846761226654 train acc 0.8125
epoch 9 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 9 test acc 0.8014112903225806


  0%|          | 0/122 [00:00<?, ?it/s]

epoch 10 batch id 1 loss 0.6103705763816833 train acc 0.8125
epoch 10 train acc 0.800637719050311


  0%|          | 0/31 [00:00<?, ?it/s]

epoch 10 test acc 0.8014112903225806


In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/quality(10)_0402.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/quality(10)_0402.h5')

### 족보

In [ ]:
dataset_train = tr_list[5]
dataset_test = te_list[5]

len(dataset_train), len(dataset_test)

(1217, 305)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 32
batch_size = 32
warmup_ratio = 0.1
num_epochs = 10
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
data_train[0]

#첫 번째는 패딩된 시퀀스, 두 번째는 길이와 타입에 대한 내용, 세 번째는 어텐션 마스크 시퀀스

(array([   2, 1042, 7828,  905, 6371, 2795, 7088, 1698, 3942, 2229, 6701,
         517,   54, 2423, 5640, 6044, 7330, 1185, 5398, 6156,  793, 7828,
        2890, 7096, 7275,  517,   54,  702, 2890,  517,  363,    3],
       dtype=int32),
 array(32, dtype=int32),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32),
 2)

In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=5)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=5)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
for e in range(num_epochs):
    train_acc = 0.0
    test_acc = 0.0
    model.train()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(train_dataloader)):
        optimizer.zero_grad()
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        loss = loss_fn(out, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        optimizer.step()
        scheduler.step()  # Update learning rate schedule
        train_acc += calc_accuracy(out, label)
        if batch_id % log_interval == 0:
            print("epoch {} batch id {} loss {} train acc {}".format(e+1, batch_id+1, loss.data.cpu().numpy(), train_acc / (batch_id+1)))
    print("epoch {} train acc {}".format(e+1, train_acc / (batch_id+1)))
    
    model.eval()
    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)
        valid_length= valid_length
        label = label.long().to(device)
        out = model(token_ids, valid_length, segment_ids)
        test_acc += calc_accuracy(out, label)
    print("epoch {} test acc {}".format(e+1, test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  """


  0%|          | 0/39 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 5 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


epoch 1 batch id 1 loss 1.118022084236145 train acc 0.375
epoch 1 train acc 0.422275641025641


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 1 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 2 batch id 1 loss 0.9685918688774109 train acc 0.53125
epoch 2 train acc 0.4463141025641026


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 2 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 3 batch id 1 loss 0.9721447229385376 train acc 0.53125
epoch 3 train acc 0.44551282051282054


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 3 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 4 batch id 1 loss 0.9635517001152039 train acc 0.5625
epoch 4 train acc 0.44871794871794873


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 4 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 5 batch id 1 loss 0.9286297559738159 train acc 0.59375
epoch 5 train acc 0.43189102564102566


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 5 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 6 batch id 1 loss 0.9842881560325623 train acc 0.46875
epoch 6 train acc 0.41025641025641024


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 6 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 7 batch id 1 loss 0.9556646347045898 train acc 0.46875
epoch 7 train acc 0.42628205128205127


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 7 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 8 batch id 1 loss 0.9852082133293152 train acc 0.4375
epoch 8 train acc 0.43990384615384615


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 8 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 9 batch id 1 loss 0.9610539674758911 train acc 0.5
epoch 9 train acc 0.42467948717948717


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 9 test acc 0.45441176470588235


  0%|          | 0/39 [00:00<?, ?it/s]

epoch 10 batch id 1 loss 0.9869190454483032 train acc 0.46875
epoch 10 train acc 0.44310897435897434


  0%|          | 0/10 [00:00<?, ?it/s]

epoch 10 test acc 0.45441176470588235


In [ ]:
torch.save(model, '/content/drive/Shareddrives/NLP모델링/상윤/jokbo(10)_0402.h5')

saved_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/jokbo(10)_0402.h5')

* epoch 3 성적 0.71971
* epoch 3 로드 test acc 0.6358735380116959
* epoch 3 강의력 test acc 0.6191013313609468
* epoch 3 교수님 성격 test acc 0.7636424731182796
*epoch 10 강의질 test acc 0.8014112903225806
* epoch 10 족보 test acc 0.45441176470588235

# Predict

In [ ]:
score_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/score(10)_0404.h5')
load_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/load(8)_0404.h5')
teaching_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/teaching(8)_0404.h5')
teacher_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/teacher(8)_0404.h5')
quality_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/quality(10)_0402.h5')
jokbo_model = torch.load('/content/drive/Shareddrives/NLP모델링/상윤/jokbo(10)_0402.h5')

In [ ]:
# Setting parameters
max_len = 64
batch_size = 64
warmup_ratio = 0.1
num_epochs = 10
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

def predict(predict_sentence):

    data = [predict_sentence, '0']
    dataset_another = [data]

    another_test = BERTDataset(dataset_another, 0, 1, tok, max_len, True, False)
    test_dataloader = torch.utils.data.DataLoader(another_test, batch_size=batch_size, num_workers=3)
    
    score_model.eval()
    load_model.eval()
    teaching_model.eval()
    teacher_model.eval()
    quality_model.eval()
    jokbo_model.eval()

    

    for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(test_dataloader):
        token_ids = token_ids.long().to(device)
        segment_ids = segment_ids.long().to(device)

        valid_length= valid_length
        label = label.long().to(device)

        score_out = score_model(token_ids, valid_length, segment_ids)
        load_out = load_model(token_ids, valid_length, segment_ids)
        teaching_out = teaching_model(token_ids, valid_length, segment_ids)
        teacher_out = teacher_model(token_ids, valid_length, segment_ids)
        quality_out = quality_model(token_ids, valid_length, segment_ids)
        jokbo_out = jokbo_model(token_ids, valid_length, segment_ids)


        test_eval=[]
        string = ['성적','로드','강의력','교수님성향','강의자료질','족보여부']
        for i, out in enumerate([score_out, load_out, teaching_out, teacher_out, quality_out, jokbo_out]):
          logits=out
          logits = logits.detach().cpu().numpy()
          print(string[i],': ',np.argmax(logits))

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
predict("기말에 커버해야 할 내용이 진짜 좀 많았어서 힘들었는데, 생각해보면 중간고사 안 봤었고 대신이었던 발표과제도 까다롭지 않았음. 특히 이번 시험이 좀 쉬웠던 것 같다. 근데 계속 점수로 압박 주는 거 스트레스였음")

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


성적 :  2
로드 :  1
강의력 :  1
교수님성향 :  2
강의자료질 :  1
족보여부 :  2


복수전공자로 처음 실건 수업을 듣게 되었는데, 정말 정말 만족했습니다! 영강이라 걱정했는데 거의 한국어로 해주셔서 문제 없었고, 설명을 정말 꼼꼼히 잘 해주셔서 내용 이해도 쉬웠습니다. 특히 김석경 교수님.. 성격 진짜 좋으십니다김석경 교수님 수업에서는 간단한 시험 3?번+중간 레포트 있었고, 이진국 교수님 수업에서는 매주 과제 (1시간 분량)+기말 레포트 있었습니다. 기말 레포트보다는 중간 레포트가 더 많은 시간이 소요됐었어요,, 이진국 교수님 매주 과제는 부담이 될랑 말랑 했지만 또 간단한 것들이라 충분히 할만 했고 점수도 후하게 잘 주셨습니다. 많은 거 배울 수 있어서 좋았습니다! 최고!

기말에 커버해야 할 내용이 진짜 좀 많았어서 힘들었는데, 생각해보면 중간고사 안 봤었고 대신이었던 발표과제도 까다롭지 않았음. 특히 이번 시험이 좀 쉬웠던 것 같다. 근데 계속 점수로 압박 주는 거 스트레스였음

진짜 절대 듣지마셈. 20년2학기가 어떻게 된지 모르겠지만, 학생들 일주일의 생활을 절대 생각안함. 자기 수업만 수업인줄 암. 과제가 벅차고 코로나시기에 학생들 대면활동과제로 내고 비대면에 대한 방안을 학생들이 직접 생각하게 만듬. 과로때문에 죽을 지경. 좀 반성해라 교수야. 학생들 죽일셈이냐? 비대면으로 조 활동하는데 매일 회의를 거쳐야하고 지치기 일수임. 이번 강의하고 절대 다신 강의열지마라

이승종 교수님 강의는 호불호가 크게 갈리는 것 같은데 나는 압도적 호이다.  씨엘씨는 전필이 없기 때문에... 좋게 말하면 자유롭게 나쁘게 말하면 근본이 없어서 이렇게 차근차근 철학 텍스트를 한줄 한줄 함께 읽는 수업이 정말 소중하다.... 그리고 등록금 내고 학생발표 듣는 게 아니라 교수 설명만 오롯이 들을 수 있는 수업은 더더욱 소중하다. 교수님께서 수업시간에 다른 일을 하거나 결석을 해도 마음에 담아두지 않는 분이라 편안했고 (급한 일은 늘 생기니까요) 늘 그러시듯이 학생에게 녹취를 풀어서 올리게 하고 태도점수에 반영하시기 때문에 결석해도 따라잡기가 수월했다. 교수님 강의 자체가 좋기 때문에 녹취를 푸는 것도 좋았다.  교수님이 아는 것이 정말 많고, 내용 정리정돈이 잘 되어있고, 즉 수업 준비를 정말 잘 해오시고, 따라서 설명을 정말 잘 하시기 때문에 다방면으로 얻은 것이 정말 많다.  다만 교수님께서 말투가 정말 조곤조곤하시고, 텍스트에 한시간이든 두시간이든 집중하는 게 쉬운 일이 아니라서 정신줄을 잘 잡아야 한다.  나는 노트북으로 필기를 하다가 자꾸 딴 길로 새서 나중에 리딩을 뽑아 손필기하는 것이 도움이 많이 됐다.

교수님도 친절하시고 학점도 잘 주시고 처음에 툴 다루는 게 어려워도 정말 자세히 알려주심 조교님도 메일 잘 받아주시고 천사였음 팀플과 매주 과제가 꽤 부담되긴 하지만 그만큼 툴 익히는데 많이 도움 됐고 하다보면 꽤 재밌음

In [ ]:
#질문 무한반복하기! 0 입력시 종료
end = 1
while end == 1 :
    sentence = input("하고싶은 말을 입력해주세요 : ")
    if sentence == '0' :
        break
    predict(sentence)
    print("\n")

하고싶은 말을 입력해주세요 : 수업에서 얻어가는건 별로 없습니다. 독학 열심히 하세요 교수님은 좋으신분 같은데 수업 방식이 너무 빡셉니다. 학점은 잘 주시는 편이세요 열심히 하면 학점은 잘 받을 수 있습니다. 시험도 빡세고 과제도 많습니다. 교수님 질문 하면 잘 받아주세요 족보는 안타는 것 같고 강의 자료는 매우 좋은 편입니다. 혼자 복습하실 때 도움 많이 되실꺼에요


/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


성적 :  1
로드 :  0
강의력 :  0
교수님성향 :  2
강의자료질 :  1
족보여부 :  2


하고싶은 말을 입력해주세요 : 과제 많고 조별과제 있고 시험도 있고 참여도도 보지만 저는 높은 악명을 들어서인지 괜찮았습니다! 교수님이 하라는대로만 하면 학점도 잘 주셔요 배워가는게 많았던 수업
성적 :  1
로드 :  0
강의력 :  1
교수님성향 :  1
강의자료질 :  1
족보여부 :  2


하고싶은 말을 입력해주세요 : 진짜 절대 듣지마셈. 20년2학기가 어떻게 된지 모르겠지만, 학생들 일주일의 생활을 절대 생각안함. 자기 수업만 수업인줄 암. 과제가 벅차고 코로나시기에 학생들 대면활동과제로 내고 비대면에 대한 방안을 학생들이 직접 생각하게 만듬. 과로때문에 죽을 지경. 좀 반성해라 교수야. 학생들 죽일셈이냐? 비대면으로 조 활동하는데 매일 회의를 거쳐야하고 지치기 일수임. 이번 강의하고 절대 다신 강의열지마라 학점도 엄청 짜게 주심
성적 :  2
로드 :  0
강의력 :  1
교수님성향 :  0
강의자료질 :  1
족보여부 :  2


하고싶은 말을 입력해주세요 : 진짜 절대 듣지마셈. 20년2학기가 어떻게 된지 모르겠지만, 학생들 일주일의 생활을 절대 생각안함. 자기 수업만 수업인줄 암. 과제가 벅차고 코로나시기에 학생들 대면활동과제로 내고 비대면에 대한 방안을 학생들이 직접 생각하게 만듬. 과로때문에 죽을 지경. 좀 반성해라 교수야. 학생들 죽일셈이냐? 비대면으로 조 활동하는데 매일 회의를 거쳐야하고 지치기 일수임. 이번 강의하고 절대 다신 강의열지마라 학점은 엄청 잘 주심
성적 :  2
로드 :  0
강의력 :  1
교수님성향 :  0
강의자료질 :  1
족보여부 :  2


하고싶은 말을 입력해주세요 : 복수전공자로 처음 실건 수업을 듣게 되었는데, 정말 정말 만족했습니다! 영강이라 걱정했는데 거의 한국어로 해주셔서 문제 없었고, 설명을 정말 꼼꼼히 잘 해주셔서 내용 이해도 쉬웠습니다. 특히 김석경 교수님.. 성격 진짜 좋으십니다김석경 교

* 0을 잘 뽑아내지 못한다...  

## F-1 score

### 성적

In [ ]:
dataset_train = tr_list[0]
dataset_test = te_list[0]

len(dataset_train), len(dataset_test)

(14937, 3735)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#BERT 모델 불러오기
model = BERTClassifier(bertmodel,  dr_rate=0.5).to(device)

#optimizer와 schedule 설정
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]

optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate)
loss_fn = nn.CrossEntropyLoss()

t_total = len(train_dataloader) * num_epochs
warmup_step = int(t_total * warmup_ratio)

scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=warmup_step, num_training_steps=t_total)

#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


In [ ]:
np.append(np.array([]),np.array([2,1])).astype(int)

array([2, 1])

In [ ]:
test_acc = 0
pred = []
label_list = []

score_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = score_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/117 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.7300840765514678


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['깐깐함','보통','학점느님']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

         깐깐함       0.29      0.15      0.19       110
          보통       0.66      0.65      0.65      1401
        학점느님       0.78      0.81      0.80      2224

    accuracy                           0.73      3735
   macro avg       0.58      0.53      0.55      3735
weighted avg       0.72      0.73      0.73      3735



### 로드

In [ ]:
dataset_train = tr_list[1]
dataset_test = te_list[1]

len(dataset_train), len(dataset_test)

(14348, 3587)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


In [ ]:
test_acc = 0
pred = []
label_list = []

load_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = load_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/113 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.6670353982300885


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['빡셈','보통','널널함']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

          빡셈       0.44      0.44      0.44       635
          보통       0.75      0.74      0.74      1977
         널널함       0.64      0.66      0.65       975

    accuracy                           0.66      3587
   macro avg       0.61      0.61      0.61      3587
weighted avg       0.66      0.66      0.66      3587



### 강의력

In [ ]:
dataset_train = tr_list[2]
dataset_test = te_list[2]

len(dataset_train), len(dataset_test)

(16484, 4122)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


In [ ]:
test_acc = 0
pred = []
label_list = []

teaching_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = teaching_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/129 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.6510882528324389


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['못 가르침','보통','잘 가르침']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

       못 가르침       0.53      0.45      0.49       680
          보통       0.65      0.68      0.66      1715
       잘 가르침       0.69      0.71      0.70      1727

    accuracy                           0.65      4122
   macro avg       0.62      0.61      0.62      4122
weighted avg       0.65      0.65      0.65      4122



### 교수님 성격

In [ ]:
dataset_train = tr_list[3]
dataset_test = te_list[3]

len(dataset_train), len(dataset_test)

(11387, 2847)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 64
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


In [ ]:
test_acc = 0
pred = []
label_list = []

teacher_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = teacher_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/89 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.7727324211670894


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['악마','보통','천사']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

          악마       0.39      0.28      0.33       137
          보통       0.59      0.53      0.56       666
          천사       0.84      0.88      0.86      2044

    accuracy                           0.77      2847
   macro avg       0.61      0.57      0.58      2847
weighted avg       0.76      0.77      0.77      2847



### 교재 질

In [ ]:
dataset_train = tr_list[4]
dataset_test = te_list[4]

len(dataset_train), len(dataset_test)

(3901, 976)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 32
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


In [ ]:
test_acc = 0
pred = []
label_list = []

quality_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = quality_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/31 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.8014112903225806


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['강의안 나쁨','보통','강의안 좋음']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

      강의안 나쁨       0.00      0.00      0.00        87
          보통       0.80      1.00      0.89       781
      강의안 좋음       0.00      0.00      0.00       108

    accuracy                           0.80       976
   macro avg       0.27      0.33      0.30       976
weighted avg       0.64      0.80      0.71       976



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### 족보

In [ ]:
dataset_train = tr_list[5]
dataset_test = te_list[5]

len(dataset_train), len(dataset_test)

(1217, 305)

In [ ]:
class BERTDataset(Dataset):
    def __init__(self, dataset, sent_idx, label_idx, bert_tokenizer, max_len,
                 pad, pair):
        transform = nlp.data.BERTSentenceTransform(
            bert_tokenizer, max_seq_length=max_len, pad=pad, pair=pair)

        self.sentences = [transform([i[sent_idx]]) for i in dataset]
        self.labels = [np.int32(i[label_idx]) for i in dataset]

    def __getitem__(self, i):
        return (self.sentences[i] + (self.labels[i], ))

    def __len__(self):
        return (len(self.labels))

In [ ]:
# Setting parameters
max_len = 32
batch_size = 32
warmup_ratio = 0.1
num_epochs = 5
max_grad_norm = 1
log_interval = 200
learning_rate =  5e-5

In [ ]:
#토큰화
tokenizer = get_tokenizer()
tok = nlp.data.BERTSPTokenizer(tokenizer, vocab, lower=False)

data_train = BERTDataset(dataset_train, 0, 1, tok, max_len, True, False)
data_test = BERTDataset(dataset_test, 0, 1, tok, max_len, True, False)

using cached model. /content/.cache/kobert_news_wiki_ko_cased-1087f8699e.spiece


In [ ]:
train_dataloader = torch.utils.data.DataLoader(data_train, batch_size=batch_size, num_workers=3)
test_dataloader = torch.utils.data.DataLoader(data_test, batch_size=batch_size, num_workers=3)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [ ]:
class BERTClassifier(nn.Module):
    def __init__(self,
                 bert,
                 hidden_size = 768,
                 num_classes=3,   ##클래스 수 조정##
                 dr_rate=None,
                 params=None):
        super(BERTClassifier, self).__init__()
        self.bert = bert
        self.dr_rate = dr_rate
                 
        self.classifier = nn.Linear(hidden_size , num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
    
    def gen_attention_mask(self, token_ids, valid_length):
        attention_mask = torch.zeros_like(token_ids)
        for i, v in enumerate(valid_length):
            attention_mask[i][:v] = 1
        return attention_mask.float()

    def forward(self, token_ids, valid_length, segment_ids):
        attention_mask = self.gen_attention_mask(token_ids, valid_length)
        
        _, pooler = self.bert(input_ids = token_ids, token_type_ids = segment_ids.long(), attention_mask = attention_mask.float().to(token_ids.device))
        if self.dr_rate:
            out = self.dropout(pooler)
        return self.classifier(out)

In [ ]:
#정확도 측정을 위한 함수 정의
def calc_accuracy(X,Y):
    max_vals, max_indices = torch.max(X, 1)
    train_acc = (max_indices == Y).sum().data.cpu().numpy()/max_indices.size()[0]
    return train_acc
    
train_dataloader


In [ ]:
test_acc = 0
pred = []
label_list = []

jokbo_model.eval()

for batch_id, (token_ids, valid_length, segment_ids, label) in enumerate(tqdm_notebook(test_dataloader)):
    token_ids = token_ids.long().to(device)
    segment_ids = segment_ids.long().to(device)
    valid_length= valid_length
    label = label.long().to(device)
    out = jokbo_model(token_ids, valid_length, segment_ids)
    test_acc += calc_accuracy(out, label)
    max_vals, max_indices = torch.max(out,1)
    pred.extend(max_indices.cpu().numpy().tolist())
    label_list.extend(label.tolist())

print("test acc: {}".format(test_acc / (batch_id+1)))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  import sys


  0%|          | 0/10 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


test acc: 0.45441176470588235


In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score

target_names = ['족보 안 탐','애매함','족보 탐']

print(classification_report(label_list,pred, target_names=target_names))

              precision    recall  f1-score   support

      족보 안 탐       0.00      0.00      0.00        63
         애매함       0.00      0.00      0.00       101
        족보 탐       0.46      1.00      0.63       141

    accuracy                           0.46       305
   macro avg       0.15      0.33      0.21       305
weighted avg       0.21      0.46      0.29       305



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
